### PR TITLE
DAML-LF:  base GenMap on SortedMap

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR changes the underlying data structure of GenMap. 
Instead of insertion order map we use sorted map (using order defined in #4808)

This advances the state of #2256.

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
This needs to update the documentation in https://github.com/digital-asset/daml/blob/master/compiler/damlc/daml-stdlib-src/DA/Map.daml which still talks about insertion order.

**Comment by cocreature** [comment]
The LF specification also still talks about insertion order.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 1]
```diff
@@ -223,15 +224,15 @@ object Ordering extends scala.math.Ordering[SValue] {
             case SAny(t2, v2) =>
               compareType(t1, t2) -> ImmArray((v1, v2))
           }
-          case SContractId(_) => {
-            case SContractId(_) =>
-              throw new IllegalAccessException("Try to compare relative contract ids")
+          case SContractId(RelativeContractId(_)) => {
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala]: When are we planning to switch the default in sandbox to use absolute contract ids everywhere? I expect that having maps indexed by contract ids will be one of the major usecases for genmaps.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala [hunk 2]
```diff
@@ -1425,11 +1430,22 @@ class SBuiltinTest extends FreeSpec with Matchers with TableDrivenPropertyChecks
...
 
+  private val pkg =
+    p"""
+        module Mod {
+          variant Either A B = Left : A | Right : B ;
+        }
+
+    """
+
+  val compilePackages  =
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala]:
```suggestion
  val compiledPackages  =
```

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala [hunk 3]
```diff
@@ -1425,11 +1430,22 @@ class SBuiltinTest extends FreeSpec with Matchers with TableDrivenPropertyChecks
...
+    PureCompiledPackages(Map(defaultParserParameters.defaultPackageId -> pkg)).right.get
+
   private def eval(e: Expr): Either[SError, SValue] = {
     val machine = Speedy.Machine.fromExpr(
       expr = e,
       checkSubmitterInMaintainers = true,
-      compiledPackages = PureCompiledPackages(Map.empty).right.get,
+      compiledPackages = compilePackages,
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala]:
```suggestion
      compiledPackages = compiledPackages,
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 4]
```diff
@@ -223,15 +224,15 @@ object Ordering extends scala.math.Ordering[SValue] {
...
+            case SContractId(RelativeContractId(_)) =>
+              throw SErrorCrash("relative contract id are not comparable")
           }
           case SPAP(_, _, _) => {
             case SPAP(_, _, _) =>
-              throw new IllegalAccessException("Try to compare functions")
+              throw SErrorCrash("functions are not comparable")
           }
-        }(fallback = throw new IllegalAccessException("Try to compare unrelated type"))
+        }(fallback = throw new IllegalAccessException("try to compare unrelated type"))
```
Comment (gerolf-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala]: Maybe use `SErrorCrash` here as well?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala]: I am not sure, I will change the default in the sandbox ever.
In kvutils (and then in next gen sandbox) it is already the only possible mode. 
DAML-LF 1.9 will require the new contract ids scheme to be enable.

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SBuiltinTest.scala]: Thanks.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala]: > The LF specification also still talks about insertion order.

I will do it in parallel.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 5]
```diff
@@ -222,18 +217,21 @@ object SValue {
...
     SStruct(entryFields, args)
   }
 
   def toList(textMap: STextMap): SList = {
     val entries = SortedLookupList(textMap.textMap).toImmArray
-    SList(FrontStack(entries.map { case (k, v) => entry(k, v) }))
+    SList(FrontStack(entries.map { case (k, v) => entry(SText(k), v) }))
   }
 
+  def toList(textMap: SGenMap): SList =
```
Comment (gerolf-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: rename to just `map`.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: Good catch. Thanks.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 6]
```diff
@@ -107,9 +107,11 @@ sealed trait SValue {
...
         STextMap(value.transform((_, v) => v.mapContractId(f)))
       case SGenMap(value) =>
-        SGenMap((InsertOrdMap.empty[SGenMap.Key, SValue] /: value) {
-          case (acc, (SGenMap.Key(k), v)) => acc + (SGenMap.Key(k.mapContractId(f)) -> v)
-        })
+        SGenMap(
+          value.foldLeft(TreeMap.empty[SValue, SValue](svalue.Ordering)) {
+            case (acc, (k, v)) => acc.updated(k.mapContractId(f), v.mapContractId(f))
+          }
+        )
```
Comment (S11001001) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]:
```suggestion
        SGenMap{
          type O[_] = TreeMap[SValue, SValue]
          value.iterator.map {
            case (k, v) => (k.mapContractId(f), v.mapContractId(f))
          }.to[O]
        }
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 7]
```diff
@@ -290,9 +288,13 @@ object SValue {
...
-        SGenMap(InsertOrdMap(entries.toSeq.map {
-          case (k, v) => SGenMap.Key(fromValue(k)) -> fromValue(v)
-        }: _*))
+        SGenMap(
+          entries
+            .foldLeft(TreeMap.newBuilder[SValue, SValue](svalue.Ordering)) {
+              case (b, (k, v)) => b += (fromValue(k) -> fromValue(v))
+            }
+            .result()
+        )
```
Comment (S11001001) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: Similarly use the `iterator map to` pattern.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala [hunk 8]
```diff
@@ -151,25 +153,18 @@ object SValue {
 
   final case class STextMap(textMap: HashMap[String, SValue]) extends SValue
 
-  final case class SGenMap(genMap: InsertOrdMap[SGenMap.Key, SValue]) extends SValue
+  final case class SGenMap(genMap: TreeMap[SValue, SValue]) extends SValue
```
Comment (S11001001) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: Somewhere in here

```scala
  implicit def ` ordering`: Ordering[SValue] = svalue.Ordering
```

Then you should be able to remove all the explicit references.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala (hunk 6 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala (hunk 8 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala]: done

**Pull request merged**